### PR TITLE
First commits

### PR DESCRIPTION
**Pull request by itsmostafa**
Initialize Con-PCA Tasks utilizing Golang

## 🗣 Description ##
- Added Golang pre-commits
- Initialized repository with a base Golang project
- Updated Readme and Gitignore files
- Add [Chi Router](https://github.com/go-chi/chi) dependancy

Note: I placed the golang script in the root of the project instead of in an `src` folder due to golang project layout standards:
https://github.com/golang-standards/project-layout#src

## 💭 Motivation and context ##
To build a scalable, more robust version of the con-pca's task scheduler.

## 🧪 Testing ##
ran `go test -v ./...` or `make test`

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

- [ ] Add a tag or create a release.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -80,36 +88,6 @@ repos:
...
-      - id: isort
-  - repo: https://github.com/pre-commit/mirrors-mypy
-    rev: v0.990
-    hooks:
-      - id: mypy
-  - repo: https://github.com/asottile/pyupgrade
-    rev: v3.2.0
-    hooks:
-      - id: pyupgrade
-
```
Comment (jsf9k) [inline .pre-commit-config.yaml]: I suggest just leaving these hooks in.  They just won't trigger if there aren't any `*.py` files in the project.

**Review by jsf9k — COMMENTED** [review]
- What is the file `go.sum`?  Does it need to be included?
- That eight space tab setting is pretty insane.  A few levels of indentation and you'll be hitting the right hand margin.  Is that necessary?  Four spaces is more standard, but maybe eight spaces is a `go` thing.

**Comment by jsf9k** [comment]
I unclicked the checkbox titled "Tests have been added and/or modified to cover the changes in this PR" since I don't see any tests here.  If checkboxes do not apply they can be deleted, but they should not be checked if they are not satisfied.

File: .github/CODEOWNERS [hunk 2]
```diff
@@ -3,7 +3,7 @@
 # These owners will be the default owners for everything in the
 # repo. Unless a later match takes precedence, these owners will be
 # requested for review when someone opens a pull request.
-* @dav3r @felddy @jsf9k @mcdonnnj
+* @dav3r @felddy @itsmostafa @izzy64 @jsf9k @mcdonnnj
```
Comment (dav3r) [inline .github/CODEOWNERS]: In this case, I think you can remove all members of the Fusion Dev team since we are not involved in this code at all:
```suggestion
* @itsmostafa @izzy64
```

File: .pre-commit-config.yaml (hunk 1 above)
Comment (itsmostafa) [reply .pre-commit-config.yaml]: @jsf9k resolved with [4dd2f64](https://github.com/cisagov/con-pca-tasks/pull/1/commits/4dd2f64a71100678e6199cd94fde8c4c3bcfc99f)

File: .gitignore [hunk 3]
```diff
@@ -2,7 +2,21 @@
 # Files already tracked by Git are not affected.
 # See: https://git-scm.com/docs/gitignore
 
+# Binaries
```
Comment (dav3r) [inline .gitignore]: Maintain our standard comment formatting:
```suggestion
## Binaries ##
```

File: .gitignore [hunk 4]
```diff
@@ -2,7 +2,21 @@
...
+*.dylib
+
 ## Python ##
 __pycache__
 .mypy_cache
 .python-version
+.venv
+
+# Test binary, built with `go test -c`
+*.test
```
Comment (dav3r) [inline .gitignore]: `*.test` can be moved into the `## Binaries ##` section above.  Feel free to keep the comment line though if you want.

**Comment by itsmostafa** [comment]
> * What is the file `go.sum`?  Does it need to be included?
> * That eight space tab setting is pretty insane.  A few levels of indentation and you'll be hitting the right hand margin.  Is that necessary?  Four spaces is more standard, but maybe eight spaces is a `go` thing.

@jsf9k yes go.sum needs to be included as go is expecting it. it's just empty at the moment because I don't have any modules installed to the project yet. think of it as similar to `Pipfile.lock` in python or `package-lock.json` in JS.

Which file do you see the 8 spaces tab?

File: .gitignore [hunk 5]
```diff
@@ -2,7 +2,21 @@
 # Files already tracked by Git are not affected.
 # See: https://git-scm.com/docs/gitignore
 
+# Binaries
+*.exe
+*.exe~
+*.dll
+*.so
+*.dylib
```
Comment (dav3r) [inline .gitignore]: Sort lines alphabetically:
```suggestion
*.dll
*.dylib
*.exe
*.exe~
*.so
```

File: .gitignore [hunk 6]
```diff
@@ -2,7 +2,21 @@
...
 __pycache__
 .mypy_cache
 .python-version
+.venv
+
+# Test binary, built with `go test -c`
+*.test
+
+# Environment variables
+.env
```
Comment (dav3r) [inline .gitignore]: Please keep sections sorted alphabetically, so move this section up above and match our preferred formatting: `## Environment variables ##`

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -29,6 +29,14 @@ repos:
       - id: requirements-txt-fixer
       - id: trailing-whitespace
 
+  # Golang hooks
+  - repo: https://github.com/dnephin/pre-commit-golang
+    rev: v0.5.0
+    hooks:
+      - id: go-fmt
+      - id: no-go-testing
+      - id: go-unit-tests
```
Comment (dav3r) [inline .pre-commit-config.yaml]: Sort hooks alphabetically:
```suggestion
      - id: go-fmt
      - id: go-unit-tests
      - id: no-go-testing
```

**Comment by itsmostafa** [comment]
> 



> I unclicked the checkbox titled "Tests have been added and/or modified to cover the changes in this PR" since I don't see any tests here. If checkboxes do not apply they can be deleted, but they should not be checked if they are not satisfied.

@jsf9k i actually do have a test i added. It's in `controllers/controllers_test.go`. It's common practice in golang for test files to exist next to the source code you are testing. You would just append `_test.go` to the end of it. I'm aware its different from Python where all unit tests live in its own `tests` directory.

File: .pre-commit-config.yaml [hunk 8]
```diff
@@ -68,7 +76,7 @@ repos:
         args:
           # Indent by two spaces
           - -i
-          - '2'
+          - "2"
```
Comment (dav3r) [inline .pre-commit-config.yaml]: Avoid making unnecessary changes from the skeleton:
```suggestion
          - '2'
```

**Comment by jsf9k** [comment]
> > 
> 
> > I unclicked the checkbox titled "Tests have been added and/or modified to cover the changes in this PR" since I don't see any tests here. If checkboxes do not apply they can be deleted, but they should not be checked if they are not satisfied.
> 
> @jsf9k i actually do have a test i added. It's in `controllers/controllers_test.go`. It's common practice in golang for test files to exist next to the source code you are testing. You would just append `_test.go` to the end of it. I'm aware its different from Python where all unit tests live in its own `tests` directory.

I have no idea how I missed that.  Sorry!

**Review by dav3r — COMMENTED** [review]
I noted a few items to get you started:

**Comment by jsf9k** [comment]
> > * What is the file `go.sum`?  Does it need to be included?
> > * That eight space tab setting is pretty insane.  A few levels of indentation and you'll be hitting the right hand margin.  Is that necessary?  Four spaces is more standard, but maybe eight spaces is a `go` thing.
> 
> @jsf9k yes go.sum needs to be included as go is expecting it. it's just empty at the moment because I don't have any modules installed to the project yet. think of it as similar to `Pipfile.lock` in python or `package-lock.json` in JS.
> 
> Which file do you see the 8 spaces tab?

All the `*.go` files, as well as the `Makefile`.

**Comment by itsmostafa** [comment]
> 



> 



> > > * What is the file `go.sum`?  Does it need to be included?
> > > * That eight space tab setting is pretty insane.  A few levels of indentation and you'll be hitting the right hand margin.  Is that necessary?  Four spaces is more standard, but maybe eight spaces is a `go` thing.
> > 
> > 
> > @jsf9k yes go.sum needs to be included as go is expecting it. it's just empty at the moment because I don't have any modules installed to the project yet. think of it as similar to `Pipfile.lock` in python or `package-lock.json` in JS.
> > Which file do you see the 8 spaces tab?
> 
> All the `*.go` files, as well as the `Makefile`.

@jsf9k it looks like my local tab indentations are translating to 8 spaces in Github for some reason (They actually look like 4 in VSCode locally). I just pushed a commit where i used spaces instead in the Makefile: [9397e89](https://github.com/cisagov/con-pca-tasks/pull/1/commits/9397e896b2e9d28a6d7445fac40da4d44a09b0bb). I will continue to investigate. But yes I agree 8 is way too much.

**Comment by itsmostafa** [comment]
> 

@jsf9k 

> > 
> 
> > 
> 
> > > > * What is the file `go.sum`?  Does it need to be included?
> > > > * That eight space tab setting is pretty insane.  A few levels of indentation and you'll be hitting the right hand margin.  Is that necessary?  Four spaces is more standard, but maybe eight spaces is a `go` thing.
> > > 
> > > 
> > > @jsf9k yes go.sum needs to be included as go is expecting it. it's just empty at the moment because I don't have any modules installed to the project yet. think of it as similar to `Pipfile.lock` in python or `package-lock.json` in JS.
> > > Which file do you see the 8 spaces tab?
> > 
> > 
> > All the `*.go` files, as well as the `Makefile`.
> 
> @jsf9k it looks like my local tab indentations are translating to 8 spaces in Github for some reason (They actually look like 4 in VSCode locally). I just pushed a commit where i used spaces instead in the Makefile: [9397e89](https://github.com/cisagov/con-pca-tasks/pull/1/commits/9397e896b2e9d28a6d7445fac40da4d44a09b0bb). I will continue to investigate. But yes I agree 8 is way too much.

@jsf9k after doing more research, it looks like the default for tab size in Github is set to 8. if you append `?ts=4` to a github file URL, it would go down to 4 spaces like so: `https://github.com/cisagov/con-pca-tasks/blob/9397e896b2e9d28a6d7445fac40da4d44a09b0bb/controllers/controllers_test.go?ts=4`

As previously mentioned, locally on my code editor, it looks like 4 spaces.

I did some research and found this: https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-personal-account-settings/managing-your-tab-size-rendering-preference

I couldn't find a way to manage tab size at the repo level but you can update it at a user level. however it'll visually update *all* tab sizes in all repos that you view

**Comment by jsf9k** [comment]
> @jsf9k after doing more research, it looks like the default for tab size in Github is set to 8. if you append `?ts=4` to a github file URL, it would go down to 4 spaces like so: `https://github.com/cisagov/con-pca-tasks/blob/9397e896b2e9d28a6d7445fac40da4d44a09b0bb/controllers/controllers_test.go?ts=4`

You should be using spaces, not tab characters.  You should be able to tell your editor to interpret a tab keypress as four spaces.

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -29,6 +29,14 @@ repos:
       - id: requirements-txt-fixer
       - id: trailing-whitespace
 
+  # Golang hooks
```
Comment (mcdonnnj) [inline .pre-commit-config.yaml]: I'm not sure how the rest of my team feels but I would prefer these hooks come after the text file hooks section. The configuration generally has things grouped by type so these hooks should be adjacent to the Python hooks.

**Comment by itsmostafa** [comment]
> > @jsf9k after doing more research, it looks like the default for tab size in Github is set to 8. if you append `?ts=4` to a github file URL, it would go down to 4 spaces like so: `https://github.com/cisagov/con-pca-tasks/blob/9397e896b2e9d28a6d7445fac40da4d44a09b0bb/controllers/controllers_test.go?ts=4`
> 
> You should be using spaces, not tab characters. You should be able to tell your editor to interpret a tab keypress as four spaces.

@jsf9k I'm using the golang standard [gofmt](https://go.dev/blog/gofmt) to format my project which is strongly recommended. I tried looking for a way i can get the formatter to use spaces instead of indentations. gofmt is very opinionated and doesn't have flexible configuration. If I try to use spaces, the formatter will automatically replace them with indentations. They used to have a `-tabwidth` flag but now that has been [removed](https://github.com/golang/go/commit/014f3dcc837cb6789076cff4fccaa3bd221f823e)

After some research, tabs for indentation is the official Golang recommendation:
See [here](https://pkg.go.dev/cmd/gofmt) and [here](https://www.golinuxcloud.com/golang-tabs/)

Please let me know your thoughts and how you'd like me to move forward. Thanks!

**Comment by jsf9k** [comment]
> After some research, tabs for indentation is the official Golang recommendation: See [here](https://pkg.go.dev/cmd/gofmt) and [here](https://www.golinuxcloud.com/golang-tabs/)
> 
> Please let me know your thoughts and how you'd like me to move forward. Thanks!

Go ahead and let `gofmt` do what it wants with the Go source files, but make the `Makefile` use four spaces instead of tab characters.  (This will match what we do in our other repos.)

Interesting that Go has decided to re-ignite this conflict.

File: .github/CODEOWNERS (hunk 2 above)
Comment (itsmostafa) [reply .github/CODEOWNERS]: @dav3r resolved with [e95a201](https://github.com/cisagov/con-pca-tasks/pull/1/commits/e95a2016c52ef83e86c22f499deb2cc1dcbd8dd9)

File: .gitignore (hunk 6 above)
Comment (itsmostafa) [reply .gitignore]: @dav3r resolved with [3add000](https://github.com/cisagov/con-pca-tasks/pull/1/commits/3add000073e20deba5ce69e868236d0bde794655)

File: .gitignore (hunk 5 above)
Comment (itsmostafa) [reply .gitignore]: @dav3r resolved with [3add000](https://github.com/cisagov/con-pca-tasks/pull/1/commits/3add000073e20deba5ce69e868236d0bde794655)

File: .gitignore (hunk 4 above)
Comment (itsmostafa) [reply .gitignore]: resolved with [3add000](https://github.com/cisagov/con-pca-tasks/pull/1/commits/3add000073e20deba5ce69e868236d0bde794655)

File: .gitignore (hunk 3 above)
Comment (itsmostafa) [reply .gitignore]: resolved with [3add000](https://github.com/cisagov/con-pca-tasks/pull/1/commits/3add000073e20deba5ce69e868236d0bde794655)

File: .pre-commit-config.yaml (hunk 7 above)
Comment (itsmostafa) [reply .pre-commit-config.yaml]: resolved with [e3c2e82](https://github.com/cisagov/con-pca-tasks/pull/1/commits/e3c2e82955ab02e64b5d857403dcf2d4c18bedd8)

File: .pre-commit-config.yaml (hunk 8 above)
Comment (itsmostafa) [reply .pre-commit-config.yaml]: @dav3r resolved with [b732ed3](https://github.com/cisagov/con-pca-tasks/pull/1/commits/b732ed37a0ef693f9c2844d6e84edfd78f2fb382)

**Comment by itsmostafa** [comment]
> > After some research, tabs for indentation is the official Golang recommendation: See [here](https://pkg.go.dev/cmd/gofmt) and [here](https://www.golinuxcloud.com/golang-tabs/)
> > Please let me know your thoughts and how you'd like me to move forward. Thanks!
> 
> Go ahead and let `gofmt` do what it wants with the Go source files, but make the `Makefile` use four spaces instead of tab characters. (This will match what we do in our other repos.)
> 
> Interesting that Go has decided to re-ignite this conflict.

@jsf9k resolved with [446b38f](https://github.com/cisagov/con-pca-tasks/pull/1/commits/446b38f96c54b2249c0a9d61f95bae93f4a9f4ac)

File: .gitignore (hunk 6 above)
Comment (dav3r) [reply .gitignore]: You missed adding one of the `#` characters here - should be:
`## Environment variables ## `, not `# Environment variables ##`

File: .gitignore (hunk 4 above)
Comment (itsmostafa) [reply .gitignore]: Oops, missed it. @dav3r resolved with [8c33d7c](https://github.com/cisagov/con-pca-tasks/pull/1/commits/8c33d7c4e91f24e6457a0fa1941c94083a5e8111)

File: .gitignore (hunk 6 above)
Comment (itsmostafa) [reply .gitignore]: Oops, missed it. @dav3r resolved with [8c33d7c](https://github.com/cisagov/con-pca-tasks/pull/1/commits/8c33d7c4e91f24e6457a0fa1941c94083a5e8111)

File: .pre-commit-config.yaml (hunk 9 above)
Comment (dav3r) [reply .pre-commit-config.yaml]: I agree with you on this, @mcdonnnj.

Comment (itsmostafa) [reply .pre-commit-config.yaml]: @mcdonnnj would you prefer it this way? [a6d76db](https://github.com/cisagov/con-pca-tasks/pull/1/commits/a6d76db7278d096c99f904345a7e481c81e6dfcc)

**Review by jsf9k — APPROVED** [review]

File: README.md [hunk 10]
```diff
@@ -1,23 +1,28 @@
...
+## Description ##
 
-In many cases you will instead want to use one of the more specific
-skeleton projects derived from this one.
+This repo contains the tasks portion for Con-PCA, Continuous Phishing
+Assessment. This application is reponsible for executing scheduled tasks for
+Con-PCA at scale.
 
-## New Repositories from a Skeleton ##
+## Con-PCA Repositories ##
```
Comment (dav3r) [inline README.md]: Slight wording clarification:
```suggestion
## Related Con-PCA Repositories ##
```

File: README.md [hunk 11]
```diff
@@ -1,23 +1,28 @@
...
 
-Please see our [Project Setup guide](https://github.com/cisagov/development-guide/tree/develop/project_setup)
-for step-by-step instructions on how to start a new repository from
-a skeleton. This will save you time and effort when configuring a
-new repository!
+- [con-pca-api](https://github.com/cisagov/con-pca-api)
+- [con-pca-web](https://github.com/cisagov/con-pca-web)
+- [con-pca-cicd](https://github.com/cisagov/con-pca-cicd)
```
Comment (dav3r) [inline README.md]: Alphabetize:
```suggestion
- [con-pca-api](https://github.com/cisagov/con-pca-api)
- [con-pca-cicd](https://github.com/cisagov/con-pca-cicd)
- [con-pca-web](https://github.com/cisagov/con-pca-web)
```

File: controllers/controllers.go [hunk 12]
```diff
@@ -0,0 +1,11 @@
+package controllers
+
+import (
+	"fmt"
+	"net/http"
+)
+
+// HealthCheckHandler indicates that the server is up and running.
+func HealthCheckHandler(w http.ResponseWriter, r *http.Request) {
```
Comment (dav3r) [inline controllers/controllers.go:9]: I'm not really familiar with golang, so apologies if I'm misinterpreting this, but it looks like you are passing in a parameter (`r *http.Request`) that isn't being used.

**Review by dav3r — CHANGES_REQUESTED** [review]
I noted a few more small items here.

File: README.md (hunk 11 above)
Comment (itsmostafa) [reply README.md]: resolved with [2effac3](https://github.com/cisagov/con-pca-tasks/pull/1/commits/2effac3853410ab9abf9d2e8b35772f45d8b2ad2)

File: README.md (hunk 10 above)
Comment (itsmostafa) [reply README.md]: @dav3r resolved with [2effac3](https://github.com/cisagov/con-pca-tasks/pull/1/commits/2effac3853410ab9abf9d2e8b35772f45d8b2ad2)

File: controllers/controllers.go (hunk 12 above)
Comment (itsmostafa) [reply controllers/controllers.go:9]: @dav3r `mux.HandleFunc()` in line 12 in `main.go` requires a _handler_ function with those two parameters, `http.ResponseWriter` and `http.Request` to be passed in.

Per the go module documentation:
```go
func (*http.ServeMux).HandleFunc(pattern string, handler func(http.ResponseWriter, *http.Request))
// HandleFunc registers the handler function for the given pattern.
```
As you can see above, the type hint for _handler_ is `func(http.ResponseWriter, *http.Request)`. It will fail if i don't pass in both params.

Comment (dav3r) [reply controllers/controllers.go:9]: Got it - appreciate the info, thanks!

**Review by dav3r — APPROVED** [review]
👍 👍

**Comment by dav3r** [comment]
@itsmostafa You should modify the branch protection rules in this repo so that you cannot merge until all conversations have been resolved.  That is our standard setup.

**Comment by itsmostafa** [comment]
> @itsmostafa You should modify the branch protection rules in this repo so that you cannot merge until all conversations have been resolved. That is our standard setup.

done!

File: .pre-commit-config.yaml (hunk 9 above)
Comment (itsmostafa) [reply .pre-commit-config.yaml]: @mcdonnnj just wanted to follow up on this. Thanks!

**Review by mcdonnnj — CHANGES_REQUESTED** [review]
Please add an appropriate CodeQL configuration. You can crib the [cisagov/skeleton-python-library one](https://github.com/cisagov/skeleton-python-library/blob/develop/.github/workflows/codeql-analysis.yml) and just update the language configuration appropriately (and optionally change the `cron` configuration to avoid overlap).

File: .pre-commit-config.yaml (hunk 9 above)
Comment (mcdonnnj) [reply .pre-commit-config.yaml]: Yep that works.

**Comment by itsmostafa** [comment]
> Please add an appropriate CodeQL configuration. You can crib the [cisagov/skeleton-python-library one](https://github.com/cisagov/skeleton-python-library/blob/develop/.github/workflows/codeql-analysis.yml) and just update the language configuration appropriately (and optionally change the `cron` configuration to avoid overlap).

@mcdonnnj done!

File: README.md (hunk 10 above)
Comment (Lover2023) [reply README.md]: > Slight wording clarification:
> 
> ```suggestion
> 
> ## Related Con-PCA Repositories ##
> 
> ```

**Review by mcdonnnj — COMMENTED** [review]
Please add a `test` job to the GitHub Actions workflow. I see that there is `make test` functionality so I imagine it should be straightforward to add something similar to the `build` workflow much like in [the Python skeleton](https://github.com/cisagov/skeleton-python-library/blob/eb2605b2111bea8dcde8d2ac86c294e2f02045da/.github/workflows/build.yml#L105-L160). If you feel like this is too much please create an issue to log this as future work but I strongly recommend this work gets done in the near future to provide good guardrails for pull requests.

**Comment by itsmostafa** [comment]
> Please add a `test` job to the GitHub Actions workflow. I see that there is `make test` functionality so I imagine it should be straightforward to add something similar to the `build` workflow much like in [the Python skeleton](https://github.com/cisagov/skeleton-python-library/blob/eb2605b2111bea8dcde8d2ac86c294e2f02045da/.github/workflows/build.yml#L105-L160). If you feel like this is too much please create an issue to log this as future work but I strongly recommend this work gets done in the near future to provide good guardrails for pull requests.

@mcdonnnj resolved with [f0ce5c7](https://github.com/cisagov/con-pca-tasks/pull/1/commits/f0ce5c764ca219cdbcdd3b654d723bd6555b35bb) & [cd17276](https://github.com/cisagov/con-pca-tasks/pull/1/commits/cd172760f721a29949c0e3fef819b4daa28ace65)

File: .github/workflows/build.yml [hunk 13]
```diff
@@ -98,3 +98,23 @@ jobs:
       - name: Setup tmate debug session
         uses: mxschmitt/action-tmate@v3
         if: env.RUN_TMATE
+  test:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        go-version: ["1.18", "1.19"]
```
Comment (mcdonnnj) [inline .github/workflows/build.yml]: We prefer block-style over flow-style.

```suggestion
        go-version:
          - "1.18"
          - "1.19"
```

File: .github/workflows/build.yml [hunk 14]
```diff
@@ -98,3 +98,23 @@ jobs:
...
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up Go
```
Comment (mcdonnnj) [inline .github/workflows/build.yml]: Consistent with the `lint` job.

```suggestion
      # We need the Go version and Go cache location for the actions/cache step,
      # so the Go installation must happen before that.
      - id: setup-go
```

File: .github/workflows/build.yml [hunk 15]
```diff
@@ -98,3 +98,23 @@ jobs:
...
+        uses: actions/setup-go@v3
+        with:
+          go-version: ${{ matrix.go-version }}
+      - name: Install dependencies
```
Comment (mcdonnnj) [inline .github/workflows/build.yml:132]: Leveraging caching is always a good idea for jobs like this.

```suggestion
      - uses: actions/cache@v3
        env:
          BASE_CACHE_KEY: "${{ github.job }}-${{ runner.os }}-\
            go${{ steps.setup-go.outputs.go-version }}-"
        with:
          path: |
            ${{ steps.go-cache.outputs.dir }}
          key: "${{ env.BASE_CACHE_KEY }}\
            ${{ hashFiles('go.mod') }}-\
            ${{ hashFiles('go.sum') }}"
          restore-keys: |
            ${{ env.BASE_CACHE_KEY }}
      - name: Install dependencies
```

File: README.md [hunk 16]
```diff
@@ -1,23 +1,28 @@
-# con-pca-tasks #
+# Con-PCA Tasks #
 
 [![GitHub Build Status](https://github.com/cisagov/con-pca-tasks/workflows/build/badge.svg)](https://github.com/cisagov/con-pca-tasks/actions)
```
Comment (mcdonnnj) [inline README.md:5]:
```suggestion
[![CodeQL](https://github.com/cisagov/con-pca-tasks/workflows/CodeQL/badge.svg)](https://github.com/cisagov/con-pca-tasks/actions/workflows/codeql-analysis.yml)

```

File: README.md [hunk 17]
```diff
@@ -1,23 +1,28 @@
...
-well as [pre-commit hooks](https://pre-commit.com) and
-[GitHub Actions](https://github.com/features/actions) configurations
-appropriate for the major languages that we use.
+## Description ##
 
-In many cases you will instead want to use one of the more specific
-skeleton projects derived from this one.
+This repo contains the tasks portion for Con-PCA, Continuous Phishing
+Assessment. This application is reponsible for executing scheduled tasks for
+Con-PCA at scale.
```
Comment (mcdonnnj) [inline README.md]: Just some language suggestions.

```suggestion
This project contains the tasks portion for Continuous Phishing Campaign
Assessment (Con-PCA). This application is reponsible for executing scheduled
tasks for Con-PCA at scale.
```

File: README.md [hunk 18]
```diff
@@ -1,23 +1,28 @@
...
+
+## Commands ##
+
+Run application
+
+- `make run`
+
+Test application
+
+- `make test`
```
Comment (mcdonnnj) [inline README.md]: I like this flow a bit more as it sets up for more involved command usage possibilities. I'm not 100% sure this will correctly merge through the GitHub web UI as fair warning.

```suggestion
### Running the application ###

```console
make run
```

### Testing the application ###

```console
make test
```
```

**Review by mcdonnnj — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml (hunk 13 above)
Comment (itsmostafa) [reply .github/workflows/build.yml]: @mcdonnnj resolved with [7751db5](https://github.com/cisagov/con-pca-tasks/pull/1/commits/7751db53ea6145e573bea708aea853b28cc8aea0)

File: .github/workflows/build.yml (hunk 14 above)
Comment (itsmostafa) [reply .github/workflows/build.yml]: @mcdonnnj resolved with [7751db5](https://github.com/cisagov/con-pca-tasks/pull/1/commits/7751db53ea6145e573bea708aea853b28cc8aea0)

File: .github/workflows/build.yml (hunk 15 above)
Comment (itsmostafa) [reply .github/workflows/build.yml:132]: @mcdonnnj resolved with [7751db5](https://github.com/cisagov/con-pca-tasks/pull/1/commits/7751db53ea6145e573bea708aea853b28cc8aea0)

File: README.md (hunk 16 above)
Comment (itsmostafa) [reply README.md:5]: @mcdonnnj resolved with [6ab301d](https://github.com/cisagov/con-pca-tasks/pull/1/commits/6ab301d53b29c3ef056462a63f4656455bb9a9fa)
(sorry, I accidentally committed these files to the wrong commit. i'm aware the commit message doesn't accurately describe the changes)

File: README.md (hunk 17 above)
Comment (itsmostafa) [reply README.md]: @mcdonnnj resolved with [6ab301d](https://github.com/cisagov/con-pca-tasks/pull/1/commits/6ab301d53b29c3ef056462a63f4656455bb9a9fa)

File: README.md (hunk 18 above)
Comment (itsmostafa) [reply README.md]: @mcdonnnj resolved with [6ab301d](https://github.com/cisagov/con-pca-tasks/pull/1/commits/6ab301d53b29c3ef056462a63f4656455bb9a9fa)

File: .github/workflows/build.yml [hunk 19]
```diff
@@ -98,3 +98,55 @@ jobs:
...
+
+      # Cache go mod cache, used to speedup builds
+      - name: Go Mod Cache
+        uses: actions/cache@v2
+        with:
+          path: |
+            ${{ steps.go-cache.outputs.dir }}
+          key: "${{ env.BASE_CACHE_KEY }}\
+            ${{ hashFiles('go.mod') }}-\
+            ${{ hashFiles('go.sum') }}"
```
Comment (mcdonnnj) [inline .github/workflows/build.yml]: These are essentially duplicates of the earlier caching step and now there is nothing being done in this job.

File: .github/workflows/build.yml [hunk 20]
```diff
@@ -98,3 +98,55 @@ jobs:
...
         if: env.RUN_TMATE
+  test:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        go-version:
+          - "1.18"
+          - "1.19"
+    steps:
+      # We need the Go version and Go cache location for the actions/cache step,
```
Comment (mcdonnnj) [inline .github/workflows/build.yml:110]: It looks like this got accidentally removed in https://github.com/cisagov/con-pca-tasks/commit/7751db53ea6145e573bea708aea853b28cc8aea0 and was the source of test failures.

```suggestion
      - uses: actions/checkout@v3
      # We need the Go version and Go cache location for the actions/cache step,
```

File: .github/workflows/build.yml (hunk 19 above)
Comment (itsmostafa) [reply .github/workflows/build.yml:149]: @mcdonnnj so to clarify the initial caching step builds and tests the go application?

Comment (itsmostafa) [reply .github/workflows/build.yml]: resolved with [5b1493b](https://github.com/cisagov/con-pca-tasks/pull/1/commits/5b1493b033102587c6a7655d569ca0ce321f8f8f)

File: .github/workflows/build.yml (hunk 20 above)
Comment (itsmostafa) [reply .github/workflows/build.yml:110]: resolved with [5b1493b](https://github.com/cisagov/con-pca-tasks/pull/1/commits/5b1493b033102587c6a7655d569ca0ce321f8f8f)

Comment (mcdonnnj) [reply .github/workflows/build.yml:110]: Would you please move the `actions/checkout` step above the comment that is intended for the `actions/setup-go` step?

Comment (itsmostafa) [reply .github/workflows/build.yml:110]: @mcdonnnj resolved with [98209ca](https://github.com/cisagov/con-pca-tasks/pull/1/commits/98209ca2c2f00135c5a455a5fdc8fc71f9493b51)

**Review by mcdonnnj — APPROVED** [review]
This looks pretty solid. I think it would be helpful to make a `build` job in the GitHub Actions workflow that creates a binary and uploads that as an artifact, but that is not strictly necessary right now. Maybe make an issue to track that as future work?

**Comment by itsmostafa** [comment]
> This looks pretty solid. I think it would be helpful to make a `build` job in the GitHub Actions workflow that creates a binary and uploads that as an artifact, but that is not strictly necessary right now. Maybe make an issue to track that as future work?

Awesome, Sounds good. Yes I was planning on working on that effort in the near future. I'll create an issue.
